File: ThreadsApp/ThreadsApp/Utils/TAStorage.m
//
//  TAStorage.m
//  ThreadsApp
//
//  Created by Brooma Service on 31/10/2017.
//  Copyright © 2017 Brooma Service. All rights reserved.
//

#import "TAStorage.h"
#import "DictUtils.h"

static NSString* kTAMainFolder             = @"TAMainFolder";

static NSString* kClientsInfoFolder          = @"TAClientsInfo";
static NSString* kClientIdsFile              = @"TAClients.json";

static NSString* kClientsKey = @"clients";


@implementation TAStorage

+ (instancetype) sharedInstance {
    static id instance = nil;
    static dispatch_once_t onceSecurePredicate;
    dispatch_once(&onceSecurePredicate,^{
        instance = [self new];
    });
    [instance createClientDirectoryIfNeed];
    return instance;
}

- (void) createClientDirectoryIfNeed {
    BOOL isDir;
    NSString *directory = [self clientDirectory];
    if (!directory)
        return;
    NSFileManager *fm = [NSFileManager defaultManager];
    if (![fm fileExistsAtPath: directory isDirectory: &isDir])
        [fm createDirectoryAtPath: directory
      withIntermediateDirectories: YES
                       attributes: nil
                            error: nil];
}

#pragma mark - Folders

- (NSString *) appDirectory {
    NSString *appDirectory = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES)[0];
    return appDirectory;
}

- (NSString *) thrDirectory {
    NSString *thrDirectory = [self.appDirectory stringByAppendingPathComponent:kTAMainFolder];
    return thrDirectory;
}

- (NSString *) clientDirectory {
    return [self directoryWithFolder: kClientsInfoFolder inRoot: YES];
}

- (NSString *) directoryWithFolder: (NSString*) folder {
    return [self directoryWithFolder: folder inRoot: NO];
}

- (NSString *) directoryWithFolder: (NSString*) folder inRoot: (BOOL) inRoot {
    NSString *directory = (inRoot) ? self.thrDirectory : self.clientDirectory;
    if (!directory)
        return nil;
    NSString* directoryPath = [directory stringByAppendingPathComponent: folder];
    NSFileManager* fileManager = [NSFileManager defaultManager];
    NSError* error = nil;
    if (![fileManager createDirectoryAtPath: directoryPath
                withIntermediateDirectories: YES
                                 attributes: nil
                                      error: &error]){
        NSLog(@"Create directory error: %@", error);
    }
    return directoryPath;
}

- (void) setClients: (NSArray<Client*>*) clients {
    
    NSString *path = [self.clientDirectory stringByAppendingPathComponent: kClientIdsFile];
    
    NSMutableDictionary *params = [NSMutableDictionary dictionary];
    params[kClientsKey] = clients;
    
    NSFileManager *fileManager = [NSFileManager defaultManager];
    BOOL created = [fileManager createFileAtPath: path
                                        contents: [NSKeyedArchiver archivedDataWithRootObject:params]
                                      attributes: nil];
    NSLog(@"created = %@", @(created));
}

- (NSMutableArray<Client*>*) getClients {
    
    NSString *path = [self.clientDirectory stringByAppendingPathComponent: kClientIdsFile];
    NSDictionary *params = [self getJSONContentOfFile:path];
    
    if(params[kClientsKey]){
        return [[DictUtils getArrayFrom: params byKey: kClientsKey] mutableCopy];
    }
    return [@[] mutableCopy];
}

- (void) addClient: (Client*) newClient {
    
    NSMutableArray<Client*>* clients = [self getClients];
    [clients addObject: newClient];
    [self setClients: clients];
}

- (void) removeClient: (Client*) clientToRemove {
    
    NSMutableArray<Client*>* clients = [self getClients];
    
    for (NSInteger index =0; index < clients.count; index++) {
        Client* client = [clients objectAtIndex: index];
        if ([client.clientId isEqualToString: clientToRemove.clientId]) {
            [clients removeObject: client];
            index--;
        }
    }
    
    [self setClients: clients];
}

#pragma mark - Helpers

- (void) log {
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSError *error = nil;
    for (NSString *item in [fileManager contentsOfDirectoryAtPath:self.appDirectory error:&error]) {
        NSString *fullPath = [self.appDirectory stringByAppendingPathComponent:item];
        NSLog(@"item = %@", item);
        BOOL isDirectory;
        if ([fileManager fileExistsAtPath:fullPath isDirectory:&isDirectory]) {
            if (isDirectory) {
                for (NSString *jtem in [fileManager contentsOfDirectoryAtPath:fullPath error:&error]) {
                    NSLog(@"\tjtem = %@", jtem);
                }
            }
        }
        //NSData *data = [NSData dataWithContentsOfFile:[self.messagesDirectory stringByAppendingPathComponent:item]];
        //NSDictionary *dict = [NSKeyedUnarchiver unarchiveObjectWithData:data];
        //NSLog(@"content = %@", dict);
    }
}

- (NSMutableDictionary *) getJSONContentOfFile:(NSString *)path{
    NSData *data = [NSData dataWithContentsOfFile: path];
    if(data) {
        NSMutableDictionary *dict = [NSKeyedUnarchiver unarchiveObjectWithData: data];
        return dict;
    }
    return [NSMutableDictionary dictionary];
}

@end

File: ThreadsApp/ThreadsAppSwift/Utils/TAStorage.swift
//
//  TAStorage.swift
//  THRAppSwift-Polarbear
//
//  Created by Brooma Service on 08/05/2019.
//  Copyright © 2019 Brooma Service. All rights reserved.
//

import Foundation

class TAStorage {

private let kTAMainFolder = "TAMainFolder"
private let kClientsInfoFolder = "TAClientsInfo"
private let kClientIdsFile = "TAClients.json"
private let kClientsKey = "clients"
    
    
    static let instance = TAStorage()
    
    private init(){
        createClientDirectoryIfNeed()
    }
    
    private func createClientDirectoryIfNeed() {
        
        let directory = self.clientDirectory()

        if (!FileManager.default.fileExists(atPath: directory.absoluteString)) {
            do {
                try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true, attributes: nil)
            } catch {
                print("Create dir failed: \(error)")
            }
        }
        
    }

    func setClients(_ clients: Array<Client>) {
        
        let clientsFileUrl = self.clientDirectory().appendingPathComponent(kClientIdsFile, isDirectory: false)
        
        //When will drop ObjC TestApp support - swich to archiving with Codable interface
//        do {
//            let data = try PropertyListEncoder().encode(clients)
//            let success = NSKeyedArchiver.archiveRootObject(data, toFile: clientsFileUrl.path)
//            print("Clients Save : ", success ? "Succeeded" : "Failed")
//        } catch {
//            print("Clients Save : Failed")
//        }
        
        let params = NSMutableDictionary()
        params[kClientsKey] = clients;
        
        let success = NSKeyedArchiver.archiveRootObject(params, toFile: clientsFileUrl.path)
        print("Clients Save : ", success ? "Succeeded" : "Failed")
    }

    func getClients() -> Array<Client> {
        
        let path = self.clientDirectory().appendingPathComponent(kClientIdsFile, isDirectory: false)
        let unarchivedObj = NSKeyedUnarchiver.unarchiveObject(withFile: path.path)
        
        //When will drop ObjC TestApp support - swich to archiving with Codable interface
        // Check if was archived with Codable as Array
//        if let data = unarchivedObj as? Data {
//            do {
//                let decoder = PropertyListDecoder()
//                clients = try decoder.decode(Array<Client>.self, from: data)
//            } catch {
//                print("Retrieve Failed")
//            }
//        }
        // Fallback to NSCoding with Dict
//        else
        
        if let dict = unarchivedObj as? Dictionary<String, Any>,
            let clients = dict[kClientsKey] as? Array<Client> {
            return clients
        
        } else {
            return Array()
        }
    }

    func addClient(_ newClient: Client)  {
        
        var clients = self.getClients()
        clients.append(newClient)
        self.setClients(clients)
    }
    
    func removeClient(_ clientToRemove: Client)  {
        
        var clients = self.getClients()
        
        let indexToRemove = clients.firstIndex(of: clientToRemove)
        if (indexToRemove != nil) {
            clients.remove(at: indexToRemove!)
        }
        
        self.setClients(clients)
    }

    // MARK: - Folders

    func appDirectory() -> URL {
        let appPath = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first!
        
        return URL(fileURLWithPath: appPath)
    }
    
    func thrDirectory() -> URL {
        return self.appDirectory().appendingPathComponent(kTAMainFolder)
    }
    
    func clientDirectory() -> URL {
        return self.directory(withFolder: kClientsInfoFolder, inRoot: true)
    }
    
    func directory(withFolder: String) -> URL  {
        return self.directory(withFolder: withFolder, inRoot: false)
    }
    
    func directory(withFolder: String, inRoot: Bool) -> URL {
        
        let directory = inRoot ? self.thrDirectory() : self.clientDirectory()
        
        let directoryPath = directory.appendingPathComponent(withFolder, isDirectory: true)
        
        do {
            try FileManager.default.createDirectory(at: directoryPath, withIntermediateDirectories: true, attributes: nil)
        } catch {
            print("Create directory error: \(error)")
        }
        
        return directoryPath;
    }
    
    
    // MARK: - Helpers
    
    func logContents(of: URL) {
        
        do {
            let dirContents = try FileManager.default.contentsOfDirectory(at: of, includingPropertiesForKeys: nil, options: [])
            for item in dirContents {
                
                print("Item = \(item)")
                
                if (item.hasDirectoryPath) {
                    logContents(of: item)
                }
            }
        } catch {
            print("Could not get dir contents: \(error)")
        }
    }
}
